MultiBuild/BlueprintManager.cs

```diff
@@ -554,124 +523,67 @@ public static List<BuildPreview> Paste(Vector3 targetPos, float yaw, bool pasteI
-                    sizeDeviation = lastMaxWidth;
-
-                sprPos = new Vector2(newRelative.x, newRelative.y * sizeDeviation) + targetSpr;
-
-                Vector3 absoluteBeltPos = sprPos.SnapToGrid(belt.altitude * 1.3333333f / 2);
-
-                // belts have always 0 yaw
-                Quaternion absoluteBeltRot = Maths.SphericalRotation(absoluteBeltPos, 0f);
-
-                BuildPreview bp = BuildPreview.CreateSingle(belt.itemProto, belt.itemProto.prefabDesc, true);
-                bp.ResetInfos();
-                bp.desc = belt.itemProto.prefabDesc;
-                bp.item = belt.itemProto;
-
-                bp.lpos = absoluteBeltPos;
-                bp.lrot = absoluteBeltRot;
-                bp.outputToSlot = -1;
-                bp.outputFromSlot = 0;
-
-                bp.inputFromSlot = -1;
-                bp.inputToSlot = 1;
-
-                bp.outputOffset = 0;
-                bp.inputOffset = 0;
+                PasteBelt(i, data.copiedBelts[i], targetSpr, yaw);
+            }
 
-                Pose pose = new Pose(absoluteBeltPos, absoluteBeltRot);
+            if (pasteInserters && data.copiedInserters.Count > 16)
+            {
 
-                int objId = InserterPoses.AddOverride(pose, belt.itemProto);
+                var maxThreads = 3;
+                var runningThreads = maxThreads;
+                var next = -1;
+                ManualResetEvent done = new ManualResetEvent(false);
 
-                pastedEntities.Add(belt.originalId, new PastedEntity()
+                for (int i = 0; i < maxThreads; i++)
                 {
-                    type = EPastedType.belt,
-                    index = i,
-                    pose = pose,
-                    objId = objId,
-                    buildPreview = bp,
-                });
+                    ThreadPool.QueueUserWorkItem(_ =>
+                    {
+                        int item;
+                        try
+                        {
+                            PlayerAction_Build ab = BuildLogic.ClonePlayerAction_Build(actionBuild);
+                            while ((item = Interlocked.Increment(ref next)) < data.copiedInserters.Count)
+                            {
+                                PasteInserter(ab, item, data.copiedInserters[item], yaw);
+                            }
+                        }
+                        catch (Exception e)
+                        {
+                            Console.WriteLine(e.ToString());
+                        }
+                        finally
+                        {
+                            if (Interlocked.Decrement(ref runningThreads) == 0)
+                            {
+                                done.Set();
+                            }
+                        }
+                    });
+                }
 
-                previews.Add(bp);
+                done.WaitOne();
```

Comment:
Since you are synchronously waiting here anyway, Parallel.For might be better suited.